### ISSUE TITLE
Encountering error after being logged in

### ISSUE THREAD

**Issue by arnavsiva** [MaxxRK/firstrade-api#30]
error: 
Traceback (most recent call last):
  File "C:\Coding\Python\Stock Bot\Reverse-Arb-Bot\test.py", line 22, in <module>
  File "C:\Users\arnav\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\firstrade\account.py", line 183, in __init__
    self.account_statuses.append(account_status["data"])
KeyError: 'data'

but I just copied the code from the test.py. Is there an additional step that I am missing?

**Comment by MaxxRK** [comment]
Is 2fa enabled on your account? As in 2fa through a code generator?

Jun 4, 2024 12:05:06 PM Arnav Sivakumar ***@***.***>:

**Comment by arnavsiva** [comment]
like sending an email verification or a phone number verification, if so i think so

**Comment by MaxxRK** [comment]
I meant through a totp generator.

If you don't use a totp generator is your pin# correct?

If that is correct login on a browser and do not use the phone or email to verify click use another method and use your pin. Then try logging in with this.

I have not had time to add the other totp methods in here yet. I'm am not sure but firstrade may not let you use the pin after you use the phone or email until you use the pin again.

Jun 4, 2024 12:09:12 PM Arnav Sivakumar ***@***.***>:

**Comment by arnavsiva** [comment]
sorry lol, can you dumb it down for me?

basically when I sign in I just click username and password and thats it, it logs me in

**Comment by arnavsiva** [comment]
![image](https://github.com/MaxxRK/firstrade-api/assets/31674084/13d77d3e-c9c7-435b-a4d8-a39779545ae6)

this is what i see, how do i change it to login through another method. And yes my pin is correct

**Comment by MaxxRK** [comment]
That means you have already done it with a phone number /email you will have to delete  your cookies for firstrade in that browser most likely. Then when you login again it will ask for you to verify by email or phone. Underneath that it will say verify another way. If you do it that way it will go to the pin screen to login

Jun 4, 2024 12:24:18 PM Arnav Sivakumar ***@***.***>:

**Comment by arnavsiva** [comment]
alright I did that, and ran the code again and im still getting the same error. I even tried deleting the pkl file for the cookies, what should I do?

**Comment by MaxxRK** [comment]
At first glance I thought it was authentication it seems not to be...

If you have some familiarity with python code you can help me by:
Opening this file from the traceback `C:\Users\arnav\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\firstrade\account.py`

and underneath the for loop put print(self.account_numbers) on line 163

and on line 177 print(account_statuses). 
![image](https://github.com/MaxxRK/firstrade-api/assets/45299186/6dd87c64-a083-48bf-81bc-c184cd31562a)

If not I can send you a modified repository so we can look at this more.

**Comment by arnavsiva** [comment]
Also I changed it because I was reading it and you previously saved it as a variable called "account_status" so i changed all "account_statuses" to "account status" and now I got this. Idk if that was anything or I changed smtg I wasn't supposed to

['my account number']
{'rc': 'fetch_fail', 'msg': 'no response'}
Traceback (most recent call last):
  File "C:\Coding\Python\Stock Bot\Reverse-Arb-Bot\test.py", line 20, in <module>
    ft_accounts = account.FTAccountData(ft_ss)
  File "C:\Users\arnav\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\firstrade\account.py", line 185, in __init__
    self.account_status.append(account_status["data"])
KeyError: 'data'

**Comment by MaxxRK** [comment]
Yes you were correct to change that variable my bad. 

So it does not like the request for some reason. I do not know exactly why. I just ran that test.py script and it ran just fine.  

`['my account number']` just to be sure this had your actual account # in it? You were just blanking it out correct? Or was this exactly what it said?

**Comment by arnavsiva** [comment]
Weird, I can share with you my exact code because I haven't hardcoded any of the variables or anything, I make it get it off a .txt file which has always worked correctly. And yes my account number was just blanked out, it showed my exact account number so I just replaced it.

Maybe is my account.py code incorrect for the API? I'm not sure to why it wouldn't be working?

Could this also be potentially due to the fact that I've logged in on my phone and website before the incorrect way? I did reset the cookies on the website so im not sure. Also should I go into urls.py and replace the user-agent with my user-agent? 

Hopefully we can fix this!

**Comment by MaxxRK** [comment]
Until I asked you to change it you had the exact account.py code from a pip install correct? You could start snooping your network traffic.

-Inspect the page go to network
-Login to firstrade, go to account dropdown, click profile, then click on account status tab.
- make sure you see "https://invest.firstrade.com/cgi-bin/account_status" and "https://invest.firstrade.com/scripts/profile/margin_v2.php" in the network urls after a filter. You should see requests and responses there too. I do not think it's your user-agent since you can log in and most users have been able to use this ok. It is always a possibility because of the nature of it being reverse engineered, but I do not think that's it.  
- On the request that ends in margin_v2.php you should see json with all of your account data in it.

I think my first questions were just because I misunderstood the question. I do not really have any proof that logging in with the other method completely locks you out of being able to do it with a pin I have not tested it myself.

**Comment by MaxxRK** [comment]
Until I asked you to change it you had the exact account.py code from a pip install correct? You could start snooping your network traffic.

Login to firstrade, go to account dropdown, click profile, then click on account status tab.

Inspect the page go to network, and make sure you see "https://invest.firstrade.com/cgi-bin/account_status" and "https://invest.firstrade.com/scripts/profile/margin_v2.php" in the network urls after a filter. You should see requests and responses there too. I do not think it's your user-agent since you can log in and most users have been able to use this ok. It is always a possibility because of the nature of it being reverse engineered, but I do not think thats it.

**Comment by arnavsiva** [comment]
Yes, I hadn't changed anything from your pip install until you asked me to add the print statements. And I get it I just signed in like that and then found this API. 

ALso I can go to the first one, the account status. But the second one gives me this as the response on the website itself: 
![image](https://github.com/MaxxRK/firstrade-api/assets/31674084/a3074a09-a2cb-4ccd-b964-9d44740c7995)

**Comment by MaxxRK** [comment]
Did you go directly to that page? Or click through the buttons. 

You need inspect the page and navigate to the page yourself then look at the traffic. Going directly to the page will not work.

**Comment by arnavsiva** [comment]
Oh should I go to firstrade and then inspect and find that page? How?

**Comment by arnavsiva** [comment]
because I don't see account status one 
![image](https://github.com/MaxxRK/firstrade-api/assets/31674084/dab4b148-2b6e-4f49-ad8c-3fca2974aca0)
or the other one
![image](https://github.com/MaxxRK/firstrade-api/assets/31674084/29e73909-1bf9-4e57-9b9b-8c74430e9cde)

**Comment by MaxxRK** [comment]
Yes you have to navigate to it with the inspection open in the browser.

Login, hover over account, click on profile, click on account status tab. You should now be able to find those urls

**Comment by arnavsiva** [comment]
I followed those steps and I still go to it and double tap the thing in the network and it still gives for the margin_v2.php one the blank white screen with JSON error

**Comment by MaxxRK** [comment]
So when you navigate to it in firstrade it shows that error? Is the account fully approved? 

Is the inspection/network tab open before you navigate to this?

**Comment by arnavsiva** [comment]
the inspection/network tab is open, and I think it is, I signed up yesterday evening and they sent an email saying congrats

**Comment by arnavsiva** [comment]
This is the email conformation, if it is possibly due to the fact that the account is not funded yet because the ACH transfer is processing, then maybe. Again I'm not sure
![image](https://github.com/MaxxRK/firstrade-api/assets/31674084/d8b198c6-ffb7-4dc5-a022-2ea83571e219)

**Comment by MaxxRK** [comment]
If you navigate to what I asked you to above and it gives you an error I would say wait for the ach to complete or email firstrade support. It should not do that.

**Comment by arnavsiva** [comment]
Alright then, I've emailed their support and I will give you an update! Thanks for your help and I hope I can use this API!

**Comment by MaxxRK** [comment]
No problem. I will do my best to help you get it figured out. Or fix it if its on the api end.

**Comment by arnavsiva** [comment]
Hey update, it works now, I have no idea what changed. I literally ran in just cause I wanted to try it and now its displaying everything. I do have a quick question. When I run the code every time do I have to delete the .pkl file (the cookies file) because today I ran it and it said credentials are incorrect and then I just deleted the .pkl file and tried again and it worked?

**Comment by MaxxRK** [comment]
Awesome! Glad to hear it.

As for the pkl file:
- Please be very cautious with where you save this file and how you handle it. It is unencrypted and has your session cookies etc in it. 
- That being said. You should not have to delete your pkl file everytime it should carry over. I have not ran into this problem myself yet, but let me know if it continues to happen. I will leave this issue open for a few more days.

**Comment by arnavsiva** [comment]
Hey, everything seems to be working perfectly! I appreciate your help and your time. This API is amazing and I'm glad that I can use it. I was wondering if we can continue to chat, maybe I can grab your discord?

**Comment by arnavsiva** [comment]
hey @MaxxRK I have another problem with it. Every time I run my script and then log into firstrade later, the script fails logging into firstrade and kind says the the credentials are wrong. How can I fix that?

**Comment by MaxxRK** [comment]
My discord is `maxxrk`. Feel free to message me there.
